text/0129-video-replay-envelope.md

```diff
@@ -0,0 +1,89 @@
+- Start Date: 2024-02-06
+- RFC Type: feature
+- RFC PR: [#129](https://github.com/getsentry/rfcs/pull/129)
+- RFC Status: draft
+
+# Summary
+
+In order to capture video replays and present them to the user, we need to define/decide:
+
+- how to transport the video data to the server
+- how to integrate the video data into the RRWeb JSON format
+- how to combine multiple replay chunks to a single replay session
+
+All of these influence one another and need to be considered together in a single RFC.
+
+Note: SDK-side implementation that is currently being worked on relies on taking screenshots and encoding them to a video.
+This is based on an evaluation where a video has much smaller size than a sequence of images (**TODO fix these: factor of X for 720p video**).
+
+# Motivation
+
+We need this to to capture replays on platforms where it's not possible/feasible to produce an HTML DOM (i.e. the native format supported by RRWeb). For example: mobile apps.
+
+<!-- # Supporting Data -->
+<!-- Metrics to help support your decision (if applicable). -->
+
+# Options Considered
+
+## Using a video, with EnvelopeItem:ReplayVideo
+
+- From the SDK, we would send a new envelope with the following items: `Replay`, `ReplayVideo` and `ReplayRecording`.
+- The newly introduced item type, [`ReplayVideo`](https://github.com/getsentry/relay/blob/5fd3969e88d3eea1f2849e55b61678cac6b14e44/relay-server/src/envelope.rs#L115C5-L115C20) is used to transport the video data.
+  The envelope item would consist of a single header line (JSON), followed by a new line and the raw video data.
+  - The header should contain at least the following metadata:
+
+    ```json
+    {
+      "segment-id": 4,
+      "size": 3440,
+      "duration": 5000,
+      "encoding": "whatever",
+      "container": "whatever",
+    }
+    ```
+
+- Additionally, it would be accompanied by an item [`ReplayRecording`](https://github.com/getsentry/relay/blob/5fd3969e88d3eea1f2849e55b61678cac6b14e44/relay-server/src/envelope.rs#L113), containing a header, e.g. `{"segment_id": 12}`, followed by a new line and the RRWeb JSON.
+  - The RRWeb JSON must start with an event of type [`EventType.Custom`](https://github.com/rrweb-io/rrweb/blob/8aea5b00a4dfe5a6f59bd2ae72bb624f45e51e81/packages/types/src/index.ts#L8-L16), containing the type `video`.
```

Comment:
@billyvg from your rrweb explanation, we also need the [metaEvent](https://github.com/rrweb-io/rrweb/blob/master/packages/types/src/index.ts#L44-L51) first, right? So the recording json will look like this:
```json
 [
    {
      "type": 4,
      "data": {
        "href": "",
        "height": 1920,
        "width": 1080
      },
      "timestamp": 1681846559381
    },
    {
      "type": 5,
      "timestamp":1681846559381,
      "data": {
        "tag": "video",
        "payload": {
          "segment-id": 4,
          "size": 3440,
          "duration": 5000,
          "encoding": "whatever",
          "container": "whatever",
          "height": 1920,
          "width": 1080,
          "frame-count": 50,
          "frame-rate-type": "constant|variable",
          "frame-rate": 10,
        }
      }
    },
    // .. breadcrumbs and friends
]
```

and this is for each segment, correct?

Reply:
@romtsn For web replays, it only sends it on the initial DOM snapshot and I think when the view port changes. So only for first segment, but it won't break the player if it's sent with every segment

Reply:
> when the view port changes

ok, that could also be the case on mobile (and also orientation change). We'll actually have to finish the current segment and start a new one in that case since the video size will change. 

So we send metaEvent for first segment and whenever the view port or orientation changes, can we add this pls @vaind?

Reply:
I've updated the spec. Also added a note that the `meta` dimensions may be different than the video dimensions in case a capture is made only on a part of the view. In such cases, the video event will have non-zero top/left fields. This is relevant for Flutter where we'd be capturing whatever user wrapped in a custom widget. It doesn't have to be the top of the tree.